File: bazelrunner/src/main/java/org/jetbrains/bsp/bazel/bazelrunner/params/BazelFlag.java
package org.jetbrains.bsp.bazel.bazelrunner.params;

import io.vavr.collection.List;

public class BazelFlag {
  public static String color(boolean enabled) {
    return arg("color", enabled ? "yes" : "no");
  }

  public static String outputProto() {
    return output("proto");
  }

  public static String output(String type) {
    return arg("output", type);
  }

  public static String keepGoing() {
    return flag("keep_going");
  }

  public static String outputGroups(List<String> groups) {
    return arg("output_groups", groups.mkString(","));
  }

  public static String aspect(String name) {
    return arg("aspects", name);
  }

  private static String arg(String name, String value) {
    return String.format("--%s=%s", name, value);
  }

  private static String flag(String name) {
    return "--" + name;
  }
}

File: bazelrunner/src/main/kotlin/org/jetbrains/bsp/bazel/bazelrunner/params/BazelFlag.kt
package org.jetbrains.bsp.bazel.bazelrunner.params

import org.jetbrains.bsp.bazel.commons.Constants.NAME
import org.jetbrains.bsp.bazel.commons.Constants.VERSION

object BazelFlag {
  @JvmStatic fun color(enabled: Boolean) =
      arg("color", if (enabled) "yes" else "no")

  @JvmStatic fun keepGoing() =
      flag("keep_going")

  @JvmStatic fun outputGroups(groups: List<String>) =
      arg("output_groups", groups.joinToString(","))

  @JvmStatic fun aspect(name: String) =
      arg("aspects", name)

  @JvmStatic fun buildManualTests(): String =
          flag("build_manual_tests")

  @JvmStatic fun buildEventBinaryPathConversion(enabled: Boolean): String =
          arg("build_event_binary_file_path_conversion", enabled.toString())

  @JvmStatic fun curses(enabled: Boolean): String =
          arg("curses", if (enabled) "yes" else "no")

  @JvmStatic fun repositoryOverride(repositoryName: String, path: String): String =
    arg("override_repository", "$repositoryName=$path")

  @JvmStatic fun testOutputAll(): String =
    arg("test_output", "all")

  @JvmStatic fun experimentalGoogleLegacyApi(): String =
      flag("experimental_google_legacy_api")

  @JvmStatic fun experimentalEnableAndroidMigrationApis(): String =
    flag("experimental_enable_android_migration_apis")

  @JvmStatic fun device(device: String): String =
    arg("device", device)

  @JvmStatic fun start(startType: String): String =
    arg("start", startType)

  @JvmStatic fun testFilter(filterExpression: String): String =
          arg("test_filter", filterExpression)

  @JvmStatic fun toolTag(): String =
    arg("tool_tag", "$NAME:$VERSION")

  private fun arg(name: String, value: String) =
      String.format("--%s=%s", name, value)

  private fun flag(name: String) =
      "--$name"
}
